### PR TITLE
add support for nested shadow dom

### PR DESCRIPTION
**Pull request by Mark-Fenng**
### scenario
If a page highly depends on custom elements, there may be lots of nested shadow dom. #754 may as an example
And the current version doesn't support recording shadow doms which were attached after the full snapshot.

### what changed

1. Hijack HTMLElement.attachShadow. When we attach a shadow dom on a created element, the old version can't detect the change and the new shadow dom can't be observed.
2. If we attach a shadow dom on a created element, the serialized element data won't contain 'isShadowHost' property. So I have to add a workaround on the replay side.
3. `notInDoc` check doesn't work for nested shadow dom so I changed the checking way.
4. add recording test for nested shadow dom 

Related issues: #38, #754

### CODE REVIEW

File: packages/rrweb/src/record/shadow-dom-manager.ts [hunk 1]
```diff
@@ -30,6 +31,16 @@ export class ShadowDomManager {
     this.scrollCb = options.scrollCb;
     this.bypassOptions = options.bypassOptions;
     this.mirror = options.mirror;
+
+    // Monkey patch 'attachShadow' to observe newly added shadow doms.
+    this.originalAttachShadow = HTMLElement.prototype.attachShadow;
+    const attachShadow = this.originalAttachShadow;
+    const manager = this;
+    HTMLElement.prototype.attachShadow = function () {
```
Comment (Mark-Fenng) [inline packages/rrweb/src/record/shadow-dom-manager.ts:39]: I guess this line may potentially have the same issue as #843 if the shadow dom is inside an iframe. I will check it later

Comment (Mark-Fenng) [reply packages/rrweb/src/record/shadow-dom-manager.ts:47]: A new patch has been pushed to solve this problem.

File: packages/rrweb/src/record/shadow-dom-manager.ts [hunk 2]
```diff
@@ -30,6 +39,17 @@ export class ShadowDomManager {
...
+      const shadowRoot = attachShadow.apply(this, arguments);
+      if (this.shadowRoot)
+        manager.addShadowRoot(this.shadowRoot, this.ownerDocument);
+      return shadowRoot;
+    };
```
Comment (Juice10) [inline packages/rrweb/src/record/shadow-dom-manager.ts:52]: You could use `patch` for this: https://github.com/rrweb-io/rrweb/blob/f74b7bb28d1f63d0a8398fdd7f80a5f86b1df01c/packages/rrweb/src/utils.ts#L169-L208

File: packages/rrweb/src/record/shadow-dom-manager.ts [hunk 3]
```diff
@@ -52,4 +72,37 @@ export class ShadowDomManager {
...
+  /**
+   * Monkey patch 'attachShadow' of an IFrameElement to observe newly added shadow doms.
+   */
+  public observeAttachShadow(iframeElement: HTMLIFrameElement) {
+    if (iframeElement.contentWindow) {
+      const originalAttachShadow = (iframeElement.contentWindow as WindowWithHTMLElement)
+        .HTMLElement.prototype.attachShadow;
+      const manager = this;
+      (iframeElement.contentWindow as WindowWithHTMLElement).HTMLElement.prototype.attachShadow = function () {
+        const shadowRoot = originalAttachShadow.apply(this, arguments);
```
Comment (Juice10) [inline packages/rrweb/src/record/shadow-dom-manager.ts:85]: You could use `patch` for this as well: https://github.com/rrweb-io/rrweb/blob/f74b7bb28d1f63d0a8398fdd7f80a5f86b1df01c/packages/rrweb/src/utils.ts#L169-L208

**Review by Juice10 — APPROVED** [review]
I'm new to shadowDom, so probably not the best to judge this, but the PR looks good to me!

**Comment by Yuyz0112** [comment]
Looks this is the only way to observe the attaching behavior.

**Pull request merged**